IDE0008 doesn't account for nullable

**Version Used**: 
3.3.0-beta2-19401-05+8ed56b826700926009731ca4d3de0a4e4e652969

**Steps to Reproduce**:
Copy/paste this into new console app:
```C#
#nullable enable
class Program
{
    private static int[]? s_data;

    static void Main()
    {
        var data = s_data;
    }
}
```
then run the IDE0008 refactoring on the `var`.

**Expected Behavior**:
Produces:
```C#
#nullable enable
class Program
{
    private static int[]? s_data;

    static void Main()
    {
        int[]? data = s_data;
    }
}
```

**Actual Behavior**:
Produces:
```C#
#nullable enable
class Program
{
    private static int[]? s_data;

    static void Main()
    {
        int[] data = s_data;
    }
}
```
resulting in the warning:
```
warning CS8600: Converting null literal or possible null value to non-nullable type.
```

--- a/src/EditorFeatures/CSharpTest/CodeRefactorings/UseExplicitOrImplicitType/UseExplicitTypeRefactoringTests.cs
+++ b/src/EditorFeatures/CSharpTest/CodeRefactorings/UseExplicitOrImplicitType/UseExplicitTypeRefactoringTests.cs
@@ -293,6 +293,38 @@ class Program
 }");
         }
 
+        [Fact]
+        public async Task TestWithTopLevelNullability()
+        {
+            var code = @"
+#nullable enable
+
+class C
+{
+    private static string? s_data;
+
+    static void Main()
+    {
+        var[||] v = s_data;
+    }
+}";
+
+            var expected = @"
+#nullable enable
+
+class C
+{
+    private static string? s_data;
+
+    static void Main()
+    {
+        string? v = s_data;
+    }
+}";
+
+            await TestInRegularAndScriptWhenDiagnosticNotAppliedAsync(code, expected);
+        }
+
         private async Task TestInRegularAndScriptWhenDiagnosticNotAppliedAsync(string initialMarkup, string expectedMarkup)
         {
             // Enabled because the diagnostic is disabled

--- a/src/EditorFeatures/CSharpTest/CodeRefactorings/UseExplicitOrImplicitType/UseExplicitTypeRefactoringTests.cs
+++ b/src/EditorFeatures/CSharpTest/CodeRefactorings/UseExplicitOrImplicitType/UseExplicitTypeRefactoringTests.cs
@@ -325,6 +325,102 @@ class C
             await TestInRegularAndScriptWhenDiagnosticNotAppliedAsync(code, expected);
         }
 
+        [Fact]
+        public async Task TestWithTopLevelAndNestedArrayNullability1()
+        {
+            var code = @"
+#nullable enable
+
+class C
+{
+    private static string?[]?[,]? s_data;
+
+    static void Main()
+    {
+        var[||] v = s_data;
+    }
+}";
+
+            var expected = @"
+#nullable enable
+
+class C
+{
+    private static string?[]?[,]? s_data;
+
+    static void Main()
+    {
+        string?[]?[,]? v = s_data;
+    }
+}";
+
+            await TestInRegularAndScriptWhenDiagnosticNotAppliedAsync(code, expected);
+        }
+
+        [Fact]
+        public async Task TestWithTopLevelAndNestedArrayNullability2()
+        {
+            var code = @"
+#nullable enable
+
+class C
+{
+    private static string?[][,]? s_data;
+
+    static void Main()
+    {
+        var[||] v = s_data;
+    }
+}";
+
+            var expected = @"
+#nullable enable
+
+class C
+{
+    private static string?[][,]? s_data;
+
+    static void Main()
+    {
+        string?[][,]? v = s_data;
+    }
+}";
+
+            await TestInRegularAndScriptWhenDiagnosticNotAppliedAsync(code, expected);
+        }
+
+        [Fact]
+        public async Task TestWithTopLevelAndNestedArrayNullability3()
+        {
+            var code = @"
+#nullable enable
+
+class C
+{
+    private static string?[]?[,][,,]? s_data;
+
+    static void Main()
+    {
+        var[||] v = s_data;
+    }
+}";
+
+            var expected = @"
+#nullable enable
+
+class C
+{
+    private static string?[]?[,][,,]? s_data;
+
+    static void Main()
+    {
+        string?[]?[,][,,]? v = s_data;
+    }
+}";
+
+            await TestInRegularAndScriptWhenDiagnosticNotAppliedAsync(code, expected);
+        }
+
         private async Task TestInRegularAndScriptWhenDiagnosticNotAppliedAsync(string initialMarkup, string expectedMarkup)
         {
             // Enabled because the diagnostic is disabled

--- a/src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs
+++ b/src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs
@@ -90,7 +90,7 @@ namespace Microsoft.CodeAnalysis.CSharp.TypeStyle
 
             if (parensDesignation is null)
             {
-                var typeSymbol = semanticModel.GetTypeInfo(typeSyntax.StripRefIfNeeded()).ConvertedType;
+                var typeSymbol = semanticModel.GetTypeInfo(typeSyntax.StripRefIfNeeded()).GetConvertedTypeWithAnnotatedNullability();
 
                 // We're going to be passed through the simplifier.  Tell it to not just convert
                 // this back to var (as that would defeat the purpose of this refactoring entirely).

--- a/src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs
+++ b/src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs
@@ -62,25 +62,51 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
             {
                 ThrowIfNameOnly();
 
-                var underlyingNonArrayType = symbol.ElementType;
-                while (underlyingNonArrayType.Kind == SymbolKind.ArrayType)
+                ITypeSymbol underlyingType = symbol;
+
+                while (underlyingType is IArrayTypeSymbol innerArray)
                 {
-                    underlyingNonArrayType = ((IArrayTypeSymbol)underlyingNonArrayType).ElementType;
+                    underlyingType = innerArray.GetElementTypeWithAnnotatedNullability();
+
+                    if (underlyingType.GetNullability() == NullableAnnotation.Annotated)
+                    {
+                        // If the inner array we just moved to is also nullable, then
+                        // we must terminate the digging now so we produce the syntax for that,
+                        // and then append the ranks we passed through at the end. This is because
+                        // nullability annotations acts as a "barrier" where we won't reorder array
+                        // through. So whereas:
+                        //
+                        //     string[][,]
+                        //
+                        // is really an array of rank 1 that has an element of rank 2,
+                        //
+                        //     string[]?[,]
+                        //
+                        // is really an array of rank 2 that has nullable elements of rank 1.
+
+                        break;
+                    }
                 }
 
-                var elementTypeSyntax = underlyingNonArrayType.GenerateTypeSyntax();
+                var elementTypeSyntax = underlyingType.GenerateTypeSyntax();
                 var ranks = new List<ArrayRankSpecifierSyntax>();
 
                 var arrayType = symbol;
-                while (arrayType != null)
+                while (arrayType != null && !arrayType.Equals(underlyingType))
                 {
                     ranks.Add(SyntaxFactory.ArrayRankSpecifier(
                         SyntaxFactory.SeparatedList(Enumerable.Repeat<ExpressionSyntax>(SyntaxFactory.OmittedArraySizeExpression(), arrayType.Rank))));
 
-                    arrayType = arrayType.ElementType as IArrayTypeSymbol;
+                    arrayType = arrayType.GetElementTypeWithAnnotatedNullability() as IArrayTypeSymbol;
                 }
 
-                var arrayTypeSyntax = SyntaxFactory.ArrayType(elementTypeSyntax, ranks.ToSyntaxList());
+                TypeSyntax arrayTypeSyntax = SyntaxFactory.ArrayType(elementTypeSyntax, ranks.ToSyntaxList());
+
+                if (symbol.GetNullability() == NullableAnnotation.Annotated)
+                {
+                    arrayTypeSyntax = SyntaxFactory.NullableType(arrayTypeSyntax);
+                }
+
                 return AddInformationTo(arrayTypeSyntax, symbol);
             }
 

